src/preamble.js

```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
 }
 #endif
 
+#if ABORT_ON_WASM_EXCEPTIONS
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// Set settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+addOnInit(function () { 
+#if DISABLE_EXCEPTION_CATCHING != 1
+  // With exception catching enabled we only handle exceptions at the top level
+  // so that the exception mechanics can work uninterrupted at the inner level.
+  // Don't do this with exception catching disabled to save some code size.
+  var wrapperDepth = 0;
+#endif
+  var mainFunc = Module["asm"]["main"];
+  var instAsm = {};
+  for (var name in Module['asm']) {
+    (function(name) {
+      var original = Module['asm'][name];
+      
+      // Wrap all functions, copy the other symbols.
+      if (typeof original === 'function') {
+        instAsm[name] = function() {
+          // Don't allow this function to be called if we're aborted!
+          if (ABORT == 1) { 
+#if ASSERTIONS
+            throw "program has already aborted!";
+#else
+            throw "DEAD";	
+#endif
+          }
+          
+#if DISABLE_EXCEPTION_CATCHING != 1
+          wrapperDepth += 1;
+#endif
+          try {
+            return original.apply(null, arguments);
+          }
+          catch (e) {
+            if(
+              original == mainFunc // rethrow any exceptions from main as they are handled in callMain and include special cases
+              || ABORT === 1 // rethrow exception if abort() was called in the original function call above
+#if DISABLE_EXCEPTION_CATCHING != 1            
+              || wrapperDepth > 1 // rethrow exceptions not caught at the top level if exception catching is enabled
+#endif
+#if SUPPORT_LONGJMP
+              || e === 'longjmp' // rethrow longjmp if enabled
+#endif
+            ) {
+              throw e;
+            }
+            
+#if ASSERTIONS
+            abort("unhandled exception: " + (typeof e === 'object' && e.stack) ? e + ", " + e.stack : e);
+#else
+            abort("UNH: " + e + ", " + e.stack);
```

Comment:
we try to be minimalistic without assertions enabled, how about
```suggestion
            abort("UHE");
```

Reply:
Yeah I realized that and tried to follow the same pattern - hence the #if. I just thought having actual stack info from where it happened even in shipping builds is nice and worth the ~15 bytes when someone actually has to debug a crash from the wild. The stack trace of the abort calls is not useful, it's the one in `e` that matters.

I'll shorten it to just: abort("UNH "+e.stack) which is smaller, but still super useful.

Reply:
Settled on `abort("UHE:" + [e, e.stack])` since FireFox doesn't print the exception as part of `.stack` like Chromium does.